syntax = "proto3";

package tensor;

// Tensor Size NxCxHxW
message TensorSize {
	int32 N = 1;
	int32 C = 2;
	int32 H = 3;
	int32 w = 4;
}

// Get Tensor Size
message GetSizeRequest {
	string id = 1;
}

message GetSizeReply {
	TensorSize size = 1;
}

// Tensor NxCxHxW
message Tensor {
	Tensor size = 1;
	bytes data = 2;
}

// Set Tensor
message SetRequest {
	string id = 1;
	Tensor tensor = 2;
}

message SetReply {
	string message = 1;	
}

// Get Tensor
message GetRequest {
	string id = 1;
}

message GetReply {
	Tensor tensor = 1;
}

// Delete Tensor message
message DelRequest {
	string id = 1;
}

message DelReply {
	string message = 1;	
}

// Add tensor
message PushRequest {
	string id = 1;
	Tensor tensor = 2;
}

message PushReply {
	string message = 1;
}

// Delete tensor slice
message PopRequest {
	string id = 1;
}

message PopReply {
	string message = 1;
}

// Tensor Service
service TensorService {
	// Get tensor size
	rpc GetSize(GetSizeRequest) returns (GetSizeReply);

	// Get/Set/Del tensor
	rpc Get(GetRequest) returns (GetReply);
	rpc Set(SetRequest) returns (SetReply);
	rpc Del(DelRequest) returns (DelReply);

	// Add tensor
	rpc LPush(PushRequest) returns (PushReply);
	rpc RPush(PushRequest) returns (PushReply);

	// Delete tensor slice
	rpc LPop(PopRequest) returns (PopReply);
	rpc RPop(PopRequest) returns (PopReply);
}

